clang/lib/Sema/SemaSYCL.cpp

```diff
@@ -2553,10 +2543,157 @@ class SyclKernelDeclCreator : public SyclKernelFieldHandler {
   }
 };
 
+// This Visitor traverses the AST of the function with
+// `sycl_kernel` attribute and returns the version of “operator()()” that is
+// called by KernelFunc. There will only be one call to KernelFunc in that
+// AST because the DPC++ headers are structured such that the user’s
+// kernel function is only called once. This ensures that the correct
+// “operator()()” function call is returned, when a named function object used
+// to define a kernel has more than one “operator()()” calls defined in it. For
+// example, in the code below, 'operator()(sycl::id<1> id)' is returned based on
+// the 'parallel_for' invocation which takes a 'sycl::range<1>(16)' argument.
+//   class MyKernel {
+//    public:
+//      void operator()() const {
+//        // code
+//      }
+//
+//      [[intel::reqd_sub_group_size(4)]] void operator()(sycl::id<1> id) const
+//      {
+//        // code
+//      }
+//    };
+//
+//    int main() {
+//
+//    Q.submit([&](sycl::handler& cgh) {
+//      MyKernel kernelFunctorObject;
+//      cgh.parallel_for(sycl::range<1>(16), kernelFunctorObject);
+//    });
+//      return 0;
+//    }
+
+class KernelCallOperatorVisitor
+    : public RecursiveASTVisitor<KernelCallOperatorVisitor> {
+
+  FunctionDecl *KernelCallerFunc;
+
+public:
+  CXXMethodDecl *CallOperator = nullptr;
+  const CXXRecordDecl *KernelObj;
+
+  KernelCallOperatorVisitor(FunctionDecl *KernelCallerFunc,
+                            const CXXRecordDecl *KernelObj)
+      : KernelCallerFunc(KernelCallerFunc), KernelObj(KernelObj) {}
+
+  bool VisitCallExpr(CallExpr *CE) {
+    Decl *CalleeDecl = CE->getCalleeDecl();
+    if (isa_and_nonnull<CXXMethodDecl>(CalleeDecl)) {
+      CXXMethodDecl *MD = cast<CXXMethodDecl>(CalleeDecl);
+      if (MD->getOverloadedOperator() == OO_Call &&
+          MD->getParent() == KernelObj) {
+        CallOperator = MD;
+      }
+    }
+    return true;
+  }
+
+  CXXMethodDecl *getCallOperator() {
+    if (CallOperator)
+      return CallOperator;
+
+    TraverseDecl(KernelCallerFunc);
+    return CallOperator;
+  }
+};
+
+class ESIMDKernelDiagnostics : public SyclKernelFieldHandler {
```

Comment:
I don't quite understand the purpose of this visitor. It accepts IsESIMD parameter, yet it doesn't do anything if it is `false`. Why the parameter then?
Also, it looks like it checks size of the arguments, but there is also an old `ArgsSizeChecker` will be run for the same kernel without taking into account that the kernel is ESIMD one. This will likely lead to double diagnosing in lots of cases.
Can we just move run of old `ArgsSizeChecker` to `constructOpenCLKernel`?

Reply:
> Can we just move run of old ArgsSizeChecker to constructOpenCLKernel?

I am not sure how we would check if the `FieldChecker `and `UnionChecker `are valid before invoking `ArgsSizeChecker `from ConstructOpenCLKernel? I would think we dont run those checkers twice.
Or am I missing something here?

```
// ArgSizeChecker needs to happen after DecompMarker has completed, since it
  // cares about the decomp attributes. DecompMarker cannot run before the
  // others, since it counts on the FieldChecker to make sure it is visiting
  // valid arrays/etc. Thus, ArgSizeChecker has its own visitation.
  if (FieldChecker.isValid() && UnionChecker.isValid()) {
    Visitor.VisitRecordBases(KernelObj, ArgsSizeChecker);
    Visitor.VisitRecordFields(KernelObj, ArgsSizeChecker);
  }
```

Reply:
I think if we reached `ConstructOpenCLKernel` normally, there is an assumption that other (presumably checking ones) visitors were ok, so we just can run `ArgsSizeChecker `there.

Reply:
> I think if we reached `ConstructOpenCLKernel` normally, there is an assumption that other (presumably checking ones) visitors were ok, so we just can run `ArgsSizeChecker` there.

Thanks, that's what I thought so too. But that explicit check for Field and Union checkers were making me second guess my belief. I will just move SYCLKernelArgsSizeChecker to constructOpenCL

Reply:
>It accepts IsESIMD parameter, yet it doesn't do anything if it is false. Why the parameter then?

The reasoning behind having the new `ESIMDKernelDiagnostics` was to diagnose only ESIMD kernel arguments size and we do this only if `IsESIMD` bool is true.

If `IsESIMD` bool is false, it implies that the kernel is a regular `SYCL` kernel and its kernel arguments size were already checked as part of `SYCLKernelArgsSizeChecker`.

Reply:
> If IsESIMD bool is false, it implies that the kernel is a regular SYCL kernel and its kernel arguments size were already checked as part of SYCLKernelArgsSizeChecker.

My point is, it seems that if the kernel is not ESIMD, it doesn't make sense to run the ESIMD visitor at all.

Reply:
If we have to follow this route, we will end up something like this:

```
if (!getLangOpts().OptRecordFile.empty()) {
    Visitor.VisitRecordBases(KernelObj, argsSizeChecker, 
                             kernel_decl, kernel_body, int_header, int_footer,
                             opt_report);
    Visitor.VisitRecordFields(KernelObj, argsSizeChecker, 
                              kernel_decl, kernel_body, int_header, int_footer,
                              opt_report);
  } else if (IsSIMDKernel) {
    Visitor.VisitRecordBases(KernelObj, argsSizeChecker, esimdKernel,
                             kernel_decl, kernel_body, int_header, int_footer);
    Visitor.VisitRecordFields(KernelObj, argsSizeChecker, esimdKernel,
                              kernel_decl, kernel_body, int_header, int_footer);
  } else {
    Visitor.VisitRecordBases(KernelObj, argsSizeChecker,
                             kernel_decl, kernel_body, int_header, int_footer);
    Visitor.VisitRecordFields(KernelObj, argsSizeChecker, 
                              kernel_decl, kernel_body, int_header, int_footer);
}
```
Plus some more if's to accommodate `opt_report` generation for ESIMD kernels.

Discussed with @Fznamznon offline and she confirmed she's fine with the way the code is right now.